### DISCUSSION TITLE
In App Messaging access additional options

### DISCUSSION THREAD

**Discussion by rjsajnani** [invertase/react-native-firebase#4970] [Q&A]
Hi guys,

I would like to access the additional parameters options via in-app messaging, they mention this when I create it (Key/value pairs that will be delivered with the message to your app).

But when my device receives the in-app message I don't see these values on the console (even after clicking it).

Is there any method or function I need to access?

Do let me know.

Thanks in advance

<img width="959" alt="Screenshot 2021-03-02 at 9 50 57 AM" src="https://user-images.githubusercontent.com/17737055/109584586-cba87080-7b3c-11eb-95ef-c9fb4cc261ab.png">

**Comment by mikehardy** [comment]
Hi there! If I am not mistaken, I believe you are referring to the ability to get the data out of an in app message when the user interacts with it?

https://firebase.google.com/docs/in-app-messaging/modify-message-behavior#implement_a_displaydelegate_to_handle_card_interactions


https://firebase.google.com/docs/in-app-messaging/modify-message-behavior#use_campaign_custom_metadata

Unfortunately, while the message suppression, data collection, and manual message trigger APIs are implemented, these native firebase SDK click listeners (which are where the data would be accessed and presented to the javascript layer) have not been implemented by react-native-firebase so this data is not available at the moment

For you (or anyone) interested in implementing them, there are lots and lots of native SDK listeners implemented already so there are quite a few chunks of code to use as patterns. 

Developing for react-native-firebase is not so hard - if you follow the https://github.com/invertase/react-native-firebase/blob/master/tests/README.md for the tests app, setup takes a little bit of time, but then the code/test/debug loop is quite fast.

Implementing simply the messageClicked handler would get you what you need, and implementing one handler may not be the work of more than an hour or two?

Here is an example to pattern off of, this is how the `firebase.auth()` auth listener works - I think it is similar to how an in-app-messaging messageClicked handler would work

- type for the method call to add a listener: https://github.com/invertase/react-native-firebase/blob/f501fffbfc1baabe7fc7ed8185ad0c5be069134d/packages/auth/lib/index.d.ts#L1269-L1289
- type for the callback method you would register in the listener, something like this but would also need an enumeration type for the dismiss type and the data bundle: https://github.com/invertase/react-native-firebase/blob/f501fffbfc1baabe7fc7ed8185ad0c5be069134d/packages/auth/lib/index.d.ts#L638-L655
- example enumeration type to show how to model DismissType for the click argument https://github.com/invertase/react-native-firebase/blob/f501fffbfc1baabe7fc7ed8185ad0c5be069134d/packages/firestore/lib/index.d.ts#L188
- android listener state management example from Java
  - https://github.com/invertase/react-native-firebase/blob/f501fffbfc1baabe7fc7ed8185ad0c5be069134d/packages/auth/android/src/main/java/io/invertase/firebase/auth/ReactNativeFirebaseAuthModule.java#L80
  - https://github.com/invertase/react-native-firebase/blob/f501fffbfc1baabe7fc7ed8185ad0c5be069134d/packages/auth/android/src/main/java/io/invertase/firebase/auth/ReactNativeFirebaseAuthModule.java#L103-L115
  - https://github.com/invertase/react-native-firebase/blob/f501fffbfc1baabe7fc7ed8185ad0c5be069134d/packages/auth/android/src/main/java/io/invertase/firebase/auth/ReactNativeFirebaseAuthModule.java#L136-L184
- example of testing https://github.com/invertase/react-native-firebase/blob/f501fffbfc1baabe7fc7ed8185ad0c5be069134d/packages/auth/e2e/auth.e2e.js#L461-L503

Here is the Objective-C reference https://firebase.google.com/docs/reference/ios/firebaseinappmessaging/api/reference/Protocols/FIRInAppMessagingDisplayDelegate
Here is Java (the add listener methods are the important ones) https://firebase.google.com/docs/reference/android/com/google/firebase/inappmessaging/FirebaseInAppMessaging

**Reply by rjsajnani** [reply]
Thanks, @mikehardy will definitely check it out and try to implement it for react-native-firebase.
